### PR TITLE
listception

### PR DESCRIPTION
**Pull request by lpmi-13**
Still having issues with the lists (you like how I kept them in there so I could use my original Size method call? Yeah, me too!)

I guess I'm still a bit vague on what these CommitMessageGrader custom classes do and what they return and such. No javadocs, obvs.

### CODE REVIEW

**Comment by richdouglasevans** [comment]
progress

Almost there!

File: src/main/java/lpmi13/nlp/commit/ChrisBeamsStyleCommitMessageGrader.java [hunk 1]
```diff
@@ -28,7 +28,36 @@ public ChrisBeamsStyleCommitMessageGrader(LexicalizedParser parser) {
...
         List<TaggedWord> words = taggedWords(message);
 
-        throw new UnsupportedOperationException("Not yet implemented.");
+        Boolean resultLengthWithinBounds = processForLength(words);
+
+        Boolean firstVerbIsVB = processForFirstVerb(words);
+
+        Boolean noPunctuationEnding = processForPunctuationEnding(words);
+
+        Boolean testPass = resultLengthWithinBounds && firstVerbIsVB && noPunctuationEnding;
```
Comment (richdouglasevans) [inline src/main/java/lpmi13/nlp/commit/ChrisBeamsStyleCommitMessageGrader.java]: If you make the changes suggested by my review comments below you'll find that all but one of the tests pass. Hurrah!

The test that doesn't pass is the edge case of an empty commit message. There's a bug in `processForPunctuationEnding` that the following question should help surface: _"What if there's not even a single word?"_

There's another small issue around the way you've structured the above block of code. I like that you've made all three tests explicit. Imagine a big old barrel of apples and I've said "You win a million dollars as soon as you find an apple in that barrel"... well, as soon as you pluck one apple from the barrel you've won. There's no need to keep searching through the entire barrel of apples because the win condition has already been satisfied.

We both know that there's a bug in the `processForPunctuationEnding` 'cos I told you so above 😄 But, given my whole barrel of apples screed above, is there some way we can change the above block so that all the tests pass... without fixing the bug in `processForPunctuationEnding`?

> I'm not trying to be a dick here with this shite about barrels of apples. More trying to show you the door so that you can proceed down the path yourself. If this style grinds your gears then just let me know and I'll switch tack 😄

**Comment by richdouglasevans** [comment]
> I guess I'm still a bit vague on what these CommitMessageGrader custom classes do and what they return and such.

Sure thing. I'll flesh out the contract of the method tomorrow... off to bed 😴

**Comment by richdouglasevans** [comment]
> I guess I'm still a bit vague on what these CommitMessageGrader custom classes do and what they return and such.

When I feed a `CommitMessageGrader` a commit message, I want that commit message to be graded, and for the grade to be returned to me. I don't care _how_ it does the grading, only that it does it.

In Java, this is expressed like so:

```java
CommitMessageGrade grade(CommitMessage message)
```

Let's break that down into pieces.

`CommitMessageGrade` is the return type of the method. It's the grade that has been calculated plus the `CommitMessage` itself.

`grade` is the name of the method. In this context the word grade is a verb because when the method is invoked, grading will happen. We might have named the method `gradeThis`, `calculateTheGrade`, or `processTheCommitMessageAndReturnTheGrade`.

> I don't dig any of those alternate names because they're verbose ornamentation on the activity that's happening. What does this method _do_? It grades stuff, so let's call it `grade`.

`(CommitMessage message)` is the single argument to the `grade` method. The argument is called message and, this being Java, we specify the type of the argument: in this case it's an instance of the `CommitMessage` type.

And it'd be invoked like so:

```java
CommitMessageGrade grade = grader.grade(new CommitMessage("This is my commit message"));
```

Does that help?

I might equally have modelled the method like so:

```java
String grade(String message)
```

And it'd be invoked like so:

```java
String grade = grader.grade("This is my commit message");
```

This looks simpler -- it is -- but it's also vague. What is this grade that is returned? Is it a score between 1-10, a letter range such as A-F, or a handwavey sentence like "Well done, this looks good"? The types tell us very little.

> You may find this blog post on the subject of "string-typing" useful reading: http://blog.jessitron.com/2012/03/strong-typing-in-java-religious.html

By introducing some types we are being explicit about what goes in and what is returned. The types also open the door to us adding metadata such as the author, the SHA, the repository, and so forth to the `CommitMessage`: we can do all of that without needing to change the signature of the `CommitMessageGrader`'s method.

Yeah? We can chat about this in a 4PM slot next week if you like.

**Comment by lpmi-13** [comment]
so everything's compiling and now I just have the one test error.

It looks like the Assert(isGood()) is resolving to null or something:

isGood(lpmi13.nlp.commit.ChrisBeamsStyleCommitMessageGraderTest)  Time elapsed: 0.591 sec  <<< FAILURE!
java.lang.AssertionError: null

Anyway, I tried changing to test to assert isBad(), and that seemed to work, and I'm assuming it's not a problem with the test itself, but not super sure what to investigate here.
